llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -9542,27 +9311,21 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range,
       } else {
         Builder.insert(Recipe);
       }
-      if (Recipe->getNumDefinedValues() == 1)
+      if (Recipe->getNumDefinedValues() == 1) {
         SingleDef->replaceAllUsesWith(Recipe->getVPSingleValue());
-      else
+        // replaceAllUsesWith may invalidate the block mask cache. Update it.
+        // TODO: Include the masks as operands in the predicated VPlan directly
+        // to remove the need to keep a map of masks beyond the predication
+        // transform.
+        RecipeBuilder.updateBlockMaskCache(SingleDef,
+                                           Recipe->getVPSingleValue());
```

Comment:
One possible alternative to traversing all block masks for each such SingleDef update, is to keep another Old2New mapping, and look it up when traversing the block masks to update it once (or do two lookups when querying the cache). Plus avoid erasing Old from its parent below, and instead do so for all Olds in Old2New when no longer needed.

Reply:
Updated with an extra map, thanks